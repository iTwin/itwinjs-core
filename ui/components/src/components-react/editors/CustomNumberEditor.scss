/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "~@itwin/core-react/lib/core-react/style/variables";
@import "./variables";

.components-customnumber-editor {
  width: 100%;
  height: 100%;
  margin: 0;
}

.components-smallEditor-host {
  .components-customnumber-editor {
    &.iui-input {
      min-height: $uicore-component-height-small !important;
      padding-top: 1px !important;
      padding-bottom: 1px !important;
    }
  }
}