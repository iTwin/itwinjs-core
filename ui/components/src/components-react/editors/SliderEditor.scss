/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "~@itwin/core-react/lib/core-react/style/variables";

.components-slider-editor {
  min-width: 100%;
}

.components-slider-editor-slider {
  min-width: 100px;
  min-height: $uicore-component-height-small;
}