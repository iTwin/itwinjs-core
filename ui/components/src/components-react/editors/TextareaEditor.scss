/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "~@itwin/core-react/lib/cjs/core-react/style/colors";

.components-textarea-editor {
  min-width: 100%;
  margin: 0;
}

.components-textarea-editor-textarea {
  min-width: 150px;
  margin: 0;
}