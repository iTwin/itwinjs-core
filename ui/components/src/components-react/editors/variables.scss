/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "~@itwin/core-react/lib/core-react/inputs/variables";

$editor-iui-input-padding: $uicore-iui-input-padding;