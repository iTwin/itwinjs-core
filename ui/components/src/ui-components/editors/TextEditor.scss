/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
@import "~@bentley/ui-core/lib/ui-core/style/variables";
@import "./variables";

.components-text-editor {
  min-width:        100%;
  height:           100%;
  margin:           0;
}

.components-smallEditor-host {
  .components-text-editor {
    &.iui-input {
      min-height: $uicore-component-height-small;
      padding-top: 1px;
      padding-bottom: 1px;
    }
  }
}
