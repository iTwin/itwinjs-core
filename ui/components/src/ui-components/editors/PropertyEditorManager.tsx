/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PropertyEditors
 */

import * as React from "react";
import { TextEditor } from "./TextEditor";
import { PropertyDescription, PropertyRecord, PropertyValue, StandardTypeNames } from "@bentley/ui-abstract";
import { OutputMessageAlert, OutputMessagePriority, OutputMessageType } from "@bentley/imodeljs-frontend";

/** Asynchronous Error Message returned as part of [[AsyncValueProcessingResult]]
 * @beta
 */
export interface AsyncErrorMessage {
  priority: OutputMessagePriority;
  briefMessage: string;
  detailedMessage?: string;
  msgType?: OutputMessageType;
  alertType?: OutputMessageAlert;
  displayTime?: number;
}

/** Asynchronous Value Process Result
 * @beta
 */
export interface AsyncValueProcessingResult {
  encounteredError: boolean;
  returnValue?: PropertyValue;
  errorMessage?: AsyncErrorMessage;
}

/** DataControllers can be implemented per typename to validate and commit values.
 * @beta
 */
export interface DataController {
  validateValue(newValue: PropertyValue, record: PropertyRecord): Promise<AsyncValueProcessingResult>;
  commitValue(newValue: PropertyValue, record: PropertyRecord): Promise<AsyncValueProcessingResult>;
}

/** PropertyEditor is the base class for all property editors.
 * @beta
 */
export abstract class PropertyEditorBase implements DataController {

  public get containerHandlesBlur(): boolean {
    return true;
  }
  public get containerHandlesEscape(): boolean {
    return true;
  }
  public get containerHandlesEnter(): boolean {
    return true;
  }
  public get containerHandlesTab(): boolean {
    return true;
  }
  public customDataController: DataController | undefined = undefined;

  public abstract get reactNode(): React.ReactNode;

  public applyEditorParams(_property: PropertyDescription, _record: PropertyRecord): void { }

  public async commitValue(newValue: PropertyValue, record: PropertyRecord): Promise<AsyncValueProcessingResult> {
    if (this.customDataController)
      return this.customDataController.commitValue(newValue, record);

    return { encounteredError: false };
  }

  public async validateValue(newValue: PropertyValue, record: PropertyRecord): Promise<AsyncValueProcessingResult> {
    if (this.customDataController)
      return this.customDataController.validateValue(newValue, record);

    return { encounteredError: false };
  }

}

/** DataControllerBase is the base class for all Data Controllers.
 * @beta
 */
export abstract class DataControllerBase implements DataController {
  public async commitValue(_newValue: PropertyValue, _record: PropertyRecord): Promise<AsyncValueProcessingResult> {
    return { encounteredError: false };
  }

  public async validateValue(_newValue: PropertyValue, _record: PropertyRecord): Promise<AsyncValueProcessingResult> {
    return { encounteredError: false };
  }
}

/** Manages Property Editors. Property Editors are registered with and created by the manager.
 * @beta
 */
export class PropertyEditorManager {
  private static _editors: { [index: string]: (new () => PropertyEditorBase) } = {};
  private static _dataControllers: { [index: string]: (new () => DataControllerBase) } = {};

  public static registerEditor(editType: string, editor: new () => PropertyEditorBase, editorName?: string): void {
    const fullEditorName = PropertyEditorManager.getFullEditorName(editType, editorName);

    if (PropertyEditorManager._editors.hasOwnProperty(fullEditorName)) {
      const nameOfEditor = PropertyEditorManager._editors[fullEditorName].name;
      throw Error(`PropertyEditorManager.registerEditor error: type '${fullEditorName}' already registered to '${nameOfEditor}'`);
    }
    PropertyEditorManager._editors[fullEditorName] = editor;
  }

  private static getFullEditorName(editType: string, editorName?: string): string {
    let fullEditorName = editType;
    if (editorName)
      fullEditorName += `:${editorName}`;
    return fullEditorName;
  }

  public static registerDataController(controllerName: string, controller: new () => DataControllerBase): void {
    if (PropertyEditorManager._dataControllers.hasOwnProperty(controllerName)) {
      throw Error(`PropertyEditorManager.registerDataController error: type '${controllerName}' already registered to '${(typeof PropertyEditorManager._dataControllers[controllerName]).toString()}'`);
    }
    PropertyEditorManager._dataControllers[controllerName] = controller;
  }

  /** @internal */
  public static deregisterDataController(controllerName: string): void {
    // istanbul ignore else
    if (PropertyEditorManager._dataControllers.hasOwnProperty(controllerName)) {
      delete PropertyEditorManager._dataControllers[controllerName];
    }
  }

  public static createEditor(editType: string, editorName?: string, dataControllerName?: string): PropertyEditorBase {
    const fullEditorName = PropertyEditorManager.getFullEditorName(editType, editorName);

    let editor: PropertyEditorBase;
    if (PropertyEditorManager._editors.hasOwnProperty(fullEditorName))
      editor = new PropertyEditorManager._editors[fullEditorName]();
    else if (PropertyEditorManager._editors.hasOwnProperty(editType))
      editor = new PropertyEditorManager._editors[editType]();
    else
      editor = new BasicPropertyEditor();

    if (dataControllerName) {
      if (PropertyEditorManager._dataControllers.hasOwnProperty(dataControllerName))
        editor.customDataController = new PropertyEditorManager._dataControllers[dataControllerName]();
      else
        throw Error(`PropertyEditorManager.createEditor error: data controller '${dataControllerName}' is not registered`);
    }

    return editor;
  }

  public static hasCustomEditor(editType: string, editorName: string): boolean {
    const fullEditorName = PropertyEditorManager.getFullEditorName(editType, editorName);
    return PropertyEditorManager._editors.hasOwnProperty(fullEditorName);
  }
}

/** Basic Property Editor registered for the "text" and "string" type names.
 * It uses the [[TextEditor]] React component.
 * @beta
 */
export class BasicPropertyEditor extends PropertyEditorBase {
  public get reactNode(): React.ReactNode {
    return <TextEditor />;
  }
}

PropertyEditorManager.registerEditor(StandardTypeNames.Text, BasicPropertyEditor);
PropertyEditorManager.registerEditor(StandardTypeNames.String, BasicPropertyEditor);
