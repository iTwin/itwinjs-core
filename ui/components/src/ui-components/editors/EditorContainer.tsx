/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PropertyEditors
 */

import "./EditorContainer.scss";
import * as React from "react";
import { IModelApp, NotifyMessageDetails } from "@bentley/imodeljs-frontend";
import { PropertyRecord, PropertyValue, SpecialKey } from "@bentley/ui-abstract";
import { CommonProps } from "@bentley/ui-core";
import { AsyncErrorMessage, PropertyEditorBase, PropertyEditorManager } from "./PropertyEditorManager";

/** Arguments for the Property Updated event callback
 * @beta
 */
export interface PropertyUpdatedArgs {
  /** The property being updated. */
  propertyRecord: PropertyRecord;
  /** The new value for the property. */
  newValue: PropertyValue;
}

/** Properties for a property editor component
 * @beta
 */
export interface PropertyEditorProps extends CommonProps {
  /** The property being updated. */
  propertyRecord?: PropertyRecord;
  /** Handler for commit */
  onCommit?: (args: PropertyUpdatedArgs) => void;
  /** Handler for cancel */
  onCancel?: () => void;
  /** Handler for blur */
  onBlur?: (event: React.FocusEvent) => void;
  /** Indicates whether the Property Editor should set focus */
  setFocus?: boolean;
}

/** [[EditorContainer]] React component properties
 * @beta
 */
export interface EditorContainerProps extends CommonProps {
  /** The property being updated. */
  propertyRecord: PropertyRecord;
  /** Tooltip text */
  title?: string;
  /** Handler for commit */
  onCommit: (args: PropertyUpdatedArgs) => void;
  /** Handler for cancel */
  onCancel: () => void;
  /** Indicates whether the Property Editor should set focus */
  setFocus?: boolean;

  /** @internal */
  ignoreEditorBlur?: boolean;
}

/** @internal */
interface CloneProps extends PropertyEditorProps {
  ref: (ref: any) => void;
}

/** Interface implemented by React based type editors
 * @beta
 */
export interface TypeEditor {
  getPropertyValue: () => Promise<PropertyValue | undefined>;
  htmlElement: HTMLElement | null;
  hasFocus: boolean;
}

/**
 * EditorContainer React component used by the Table, Tree and PropertyGrid for cell editing.
 * @beta
 */
export class EditorContainer extends React.PureComponent<EditorContainerProps> {

  private _editorRef: TypeEditor | undefined;
  private _propertyEditor: PropertyEditorBase | undefined;
  private _spanRef = React.createRef<HTMLSpanElement>();

  private createEditor(): React.ReactNode {
    const editorRef = (ref: TypeEditor | undefined) => this._editorRef = ref;

    const editorProps: CloneProps = {
      ref: editorRef,
      onCommit: this._handleEditorCommit,
      onCancel: this._handleEditorCancel,
      onBlur: this._handleEditorBlur,
      propertyRecord: this.props.propertyRecord,
      setFocus: this.props.setFocus !== undefined ? this.props.setFocus : true,
      className: this.props.className,
      style: this.props.style,
    };

    const propDescription = this.props.propertyRecord.property;

    const editorName = propDescription.editor !== undefined ? propDescription.editor.name : undefined;
    this._propertyEditor = PropertyEditorManager.createEditor(propDescription.typename, editorName, propDescription.dataController);
    const editorNode: React.ReactNode = this._propertyEditor.reactNode;

    let clonedNode: React.ReactNode = null;
    // istanbul ignore else
    if (React.isValidElement(editorNode)) {
      clonedNode = React.cloneElement(editorNode, editorProps);
    }

    return clonedNode;
  }
  public componentDidMount() {
    this._spanRef.current?.addEventListener("keydown", this._handleKeyDown, true);
  }
  public componentWillUnmount() {
    // istanbul ignore next
    if (this._spanRef.current)
      this._spanRef.current.removeEventListener("keydown", this._handleKeyDown, true);
  }

  private _handleEditorBlur = (_e: React.FocusEvent) => {
    // istanbul ignore else
    if (!this.props.ignoreEditorBlur && this._propertyEditor && this._propertyEditor.containerHandlesBlur)
      this._handleContainerCommit(); // eslint-disable-line @typescript-eslint/no-floating-promises
  };

  private _handleContainerBlur = (e: React.FocusEvent) => {
    e.stopPropagation();
  };

  private _handleClick = (e: React.MouseEvent) => {
    e.stopPropagation();
  };

  private _handleKeyDown = (e: KeyboardEvent) => {
    switch (e.key) {
      case SpecialKey.Escape:
        this.onPressEscape(e);
        break;
      case SpecialKey.Enter:
        this.onPressEnter(e);
        break;
      case SpecialKey.Tab:
        this.onPressTab(e);
        break;
      default:
        e.stopPropagation();
    }
  };

  private _handleRightClick = (e: React.MouseEvent) => {
    e.stopPropagation();
  };

  private onPressEscape(_e: KeyboardEvent): void {
    // istanbul ignore else
    if (this._propertyEditor && this._propertyEditor.containerHandlesEscape) {
      this._commitCancel();
    }
  }

  private onPressEnter(e: KeyboardEvent): void {
    // istanbul ignore else
    if (this._propertyEditor && this._propertyEditor.containerHandlesEnter) {
      // istanbul ignore else
      if (this._editorRef && this._editorRef.hasFocus)
        e.stopPropagation();
      this._handleContainerCommit(); // eslint-disable-line @typescript-eslint/no-floating-promises
    }
  }

  private onPressTab(e: KeyboardEvent): void {
    // istanbul ignore else
    if (this._propertyEditor && this._propertyEditor.containerHandlesTab) {
      e.stopPropagation();
      this._handleContainerCommit(); // eslint-disable-line @typescript-eslint/no-floating-promises
    }
  }

  private displayInputFieldMessage(errorMessage: AsyncErrorMessage | undefined) {
    // istanbul ignore else
    if (errorMessage && this._editorRef) {
      const details = new NotifyMessageDetails(errorMessage.priority, errorMessage.briefMessage, errorMessage.detailedMessage);
      const htmlElement = this._editorRef && this._editorRef.htmlElement;
      // istanbul ignore else
      if (htmlElement)
        details.setInputFieldTypeDetails(htmlElement);
      if (IModelApp.notifications)
        IModelApp.notifications.outputMessage(details);
    }
  }

  private async isNewValueValid(value: PropertyValue): Promise<boolean> {
    let isValid = true;

    // istanbul ignore else
    if (this._propertyEditor && this.props.propertyRecord) {
      const validateResult = await this._propertyEditor.validateValue(value, this.props.propertyRecord);

      if (validateResult.encounteredError) {
        this.displayInputFieldMessage(validateResult.errorMessage);
        isValid = false;
      }
    } else {
      isValid = false;
    }

    return isValid;
  }

  private _handleEditorCommit = (args: PropertyUpdatedArgs): void => {
    this._commit(args); // eslint-disable-line @typescript-eslint/no-floating-promises
  };

  private _handleContainerCommit = async (): Promise<void> => {
    const newValue = this._editorRef && await this._editorRef.getPropertyValue();
    // istanbul ignore else
    if (newValue !== undefined) {
      this._commit({ propertyRecord: this.props.propertyRecord, newValue });  // eslint-disable-line @typescript-eslint/no-floating-promises
    }
  };

  private _commit = async (args: PropertyUpdatedArgs) => {
    const newValue = args.newValue;
    const isValid = await this.isNewValueValid(newValue);
    if (isValid) {
      let doCommit = true;
      // istanbul ignore else
      if (this._propertyEditor && args.propertyRecord) {
        const commitResult = await this._propertyEditor.commitValue(newValue, args.propertyRecord);
        if (commitResult.encounteredError) {
          this.displayInputFieldMessage(commitResult.errorMessage);
          doCommit = false;
        }
      }

      if (doCommit) {
        this.props.onCommit(args);
      }
    }
  };

  private _handleEditorCancel = () => {
    this._commitCancel();
  };

  private _commitCancel = () => {
    this.props.onCancel();
  };

  /** @internal */
  public render() {
    return (
      <span className="components-editor-container"
        onBlur={this._handleContainerBlur}
        onClick={this._handleClick}
        onContextMenu={this._handleRightClick}
        title={this.props.title}
        data-testid="editor-container"
        role="presentation"
        ref={this._spanRef}
      >
        {this.createEditor()}
      </span>
    );
  }
}
