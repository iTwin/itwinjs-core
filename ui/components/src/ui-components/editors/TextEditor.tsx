/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/** @packageDocumentation
 * @module PropertyEditors
 */

import "./TextEditor.scss";
import classnames from "classnames";
import * as React from "react";
import {
  IconEditorParams, InputEditorSizeParams, PrimitiveValue, PropertyEditorParams, PropertyEditorParamTypes, PropertyValue, PropertyValueFormat,
} from "@bentley/ui-abstract";
import { Icon, IconInput, Input, InputProps } from "@bentley/ui-core";
import { TypeConverterManager } from "../converters/TypeConverterManager";
import { PropertyEditorProps, TypeEditor } from "./EditorContainer";
import { UiComponents } from "../UiComponents";

/** @internal */
interface TextEditorState {
  inputValue: string;
  readonly: boolean;
  isDisabled?: boolean;
  size?: number;
  maxLength?: number;
  iconSpec?: string;
}

/** TextEditor React component that is a property editor with text input
 * @beta
 */
export class TextEditor extends React.PureComponent<PropertyEditorProps, TextEditorState> implements TypeEditor {
  private _isMounted = false;
  private _ariaLabel = UiComponents.translate("editor.text");
  private _inputElement = React.createRef<HTMLInputElement>();

  /** @internal */
  public readonly state: Readonly<TextEditorState> = {
    inputValue: "",
    readonly: false,
  };

  public async getPropertyValue(): Promise<PropertyValue | undefined> {
    const record = this.props.propertyRecord;
    let propertyValue: PropertyValue | undefined;

    // istanbul ignore else
    if (record && record.value.valueFormat === PropertyValueFormat.Primitive) {
      propertyValue = await TypeConverterManager.getConverter(record.property.typename, record.property.converter?.name).convertFromStringToPropertyValue(this.state.inputValue, record);
      (propertyValue as PrimitiveValue).displayValue = this.state.inputValue;
    }

    return propertyValue;
  }

  public get htmlElement(): HTMLElement | null {
    return this._inputElement.current;
  }

  public get hasFocus(): boolean {
    return document.activeElement === this._inputElement.current;
  }

  private _updateInputValue = (e: React.ChangeEvent<HTMLInputElement>) => {
    // istanbul ignore else
    if (this._isMounted)
      this.setState({
        inputValue: e.target.value,
      });
  };

  /** @internal */
  public componentDidMount() {
    this._isMounted = true;
    this.setStateFromProps(); // eslint-disable-line @typescript-eslint/no-floating-promises
  }

  /** @internal */
  public componentWillUnmount() {
    this._isMounted = false;
  }

  /** @internal */
  public componentDidUpdate(prevProps: PropertyEditorProps) {
    if (this.props.propertyRecord !== prevProps.propertyRecord) {
      this.setStateFromProps(); // eslint-disable-line @typescript-eslint/no-floating-promises
    }
  }

  private async setStateFromProps() {
    const record = this.props.propertyRecord;
    let initialValue = "";

    // istanbul ignore else
    if (record && record.value.valueFormat === PropertyValueFormat.Primitive) {
      const value = record.value.value;
      initialValue = await TypeConverterManager.getConverter(record.property.typename, record.property.converter?.name).convertPropertyToString(record.property, value);
    }

    const readonly = record && undefined !== record.isReadonly ? record.isReadonly : false;
    let size: number | undefined;
    let maxLength: number | undefined;
    let iconSpec: string | undefined;

    const isDisabled = record ? record.isDisabled : undefined;

    if (record && record.property && record.property.editor && record.property.editor.params) {
      const editorSizeParams = record.property.editor.params.find((param: PropertyEditorParams) => param.type === PropertyEditorParamTypes.InputEditorSize) as InputEditorSizeParams;
      if (editorSizeParams) {
        // istanbul ignore else
        if (editorSizeParams.size)
          size = editorSizeParams.size;
        // istanbul ignore else
        if (editorSizeParams.maxLength)
          maxLength = editorSizeParams.maxLength;
      }

      const iconParams = record.property.editor.params.find((param: PropertyEditorParams) => param.type === PropertyEditorParamTypes.Icon) as IconEditorParams;
      if (iconParams) {
        iconSpec = iconParams.definition.iconSpec;
      }
    }

    // istanbul ignore else
    if (this._isMounted)
      this.setState({ inputValue: initialValue, readonly, size, maxLength, isDisabled, iconSpec });
  }

  /** @internal */
  public render(): React.ReactNode {
    const className = classnames("components-cell-editor", "components-text-editor", this.props.className);
    const minSize = this.state.size ? this.state.size : 8;
    const minWidthStyle: React.CSSProperties = {
      minWidth: `${minSize * 0.75}em`,
    };
    const inputProps: InputProps = {
      type: "text",
      className,
      style: this.props.style ? this.props.style : minWidthStyle,
      readOnly: this.state.readonly,
      disabled: this.state.isDisabled,
      size: this.state.size,
      maxLength: this.state.maxLength,
      value: this.state.inputValue,
      onBlur: this.props.onBlur,
      onChange: this._updateInputValue,
      setFocus: this.props.setFocus && !this.state.isDisabled,
      ref: this._inputElement,
    };

    inputProps["aria-label"] = this._ariaLabel;

    let reactNode: React.ReactNode;
    if (this.state.iconSpec) {
      const icon = <Icon iconSpec={this.state.iconSpec} />;
      reactNode = (
        <IconInput
          {...inputProps}
          icon={icon}
          data-testid="components-text-editor"
        />
      );
    } else {
      reactNode = (
        <Input
          {...inputProps}
          data-testid="components-text-editor"
        />
      );
    }

    return reactNode;
  }
}
