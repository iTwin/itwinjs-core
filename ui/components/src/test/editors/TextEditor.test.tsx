/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

import { expect } from "chai";
import { mount, shallow } from "enzyme";
import sinon from "sinon";
import { fireEvent, render } from "@testing-library/react";
import * as React from "react";
import { IconEditorParams, InputEditorSizeParams, PropertyConverterInfo, PropertyEditorInfo, PropertyEditorParamTypes,
  PropertyRecord, PropertyValue, SpecialKey,
} from "@bentley/ui-abstract";
import { MockRender, OutputMessagePriority } from "@bentley/imodeljs-frontend";
import { TextEditor } from "../../ui-components/editors/TextEditor";
import TestUtils from "../TestUtils";
import { EditorContainer, PropertyUpdatedArgs } from "../../ui-components/editors/EditorContainer";
import { AsyncValueProcessingResult, DataControllerBase, PropertyEditorManager } from "../../ui-components/editors/PropertyEditorManager";

describe("<TextEditor />", () => {
  it("should render", () => {
    mount(<TextEditor />);
  });

  it("renders correctly", () => {
    shallow(<TextEditor />).should.matchSnapshot();
  });

  it("renders correctly with style", () => {
    shallow(<TextEditor style={{ color: "red" }} />).should.matchSnapshot();
  });

  it("getValue returns proper value after componentDidMount & setState", async () => {
    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue");
    const wrapper = mount(<TextEditor propertyRecord={record} />);

    await TestUtils.flushAsyncOperations();
    const editor = wrapper.instance() as TextEditor;
    expect(editor.state.inputValue).to.equal("MyValue");

    wrapper.unmount();
  });

  it("should support record.property.converter?.name", async () => {
    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue");
    const convertInfo: PropertyConverterInfo = { name: "" };
    record.property.converter = convertInfo;

    const wrapper = mount(<TextEditor propertyRecord={record} />);

    await TestUtils.flushAsyncOperations();
    const editor = wrapper.instance() as TextEditor;
    expect(editor.state.inputValue).to.equal("MyValue");

    wrapper.unmount();
  });

  it("HTML input onChange updates value", () => {
    const record = TestUtils.createPrimitiveStringProperty("Test1", "MyValue");
    const wrapper = mount(<TextEditor propertyRecord={record} />);
    const editor = wrapper.instance() as TextEditor;
    const inputNode = wrapper.find("input");

    expect(inputNode.length).to.eq(1);
    if (inputNode) {
      const testValue = "My new value";
      inputNode.simulate("change", { target: { value: testValue } });
      wrapper.update();
      expect(editor.state.inputValue).to.equal(testValue);
    }
  });

  it("componentDidUpdate updates the value", async () => {
    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue");
    const wrapper = mount(<TextEditor propertyRecord={record} />);

    await TestUtils.flushAsyncOperations();
    const editor = wrapper.instance() as TextEditor;
    expect(editor.state.inputValue).to.equal("MyValue");

    const testValue = "MyNewValue";
    const newRecord = TestUtils.createPrimitiveStringProperty("Test", testValue);
    wrapper.setProps({ propertyRecord: newRecord });
    await TestUtils.flushAsyncOperations();
    expect(editor.state.inputValue).to.equal(testValue);

    wrapper.unmount();
  });

  it("should support InputEditorSize params", async () => {
    const size = 4;
    const maxLength = 60;
    const editorInfo: PropertyEditorInfo = {
      params: [
        {
          type: PropertyEditorParamTypes.InputEditorSize,
          size,
          maxLength,
        } as InputEditorSizeParams,
      ],
    };

    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue", "Test", editorInfo);
    const wrapper = mount(<TextEditor propertyRecord={record} />);
    await TestUtils.flushAsyncOperations();

    const textEditor = wrapper.find(TextEditor);
    expect(textEditor.length).to.eq(1);
    expect(textEditor.state("size")).to.eq(size);
    expect(textEditor.state("maxLength")).to.eq(maxLength);

    wrapper.unmount();
  });

  it("should support IconEditor params", async () => {
    const iconSpec = "icon-placeholder";
    const editorInfo: PropertyEditorInfo = {
      params: [
        {
          type: PropertyEditorParamTypes.Icon,
          definition: { iconSpec },
        } as IconEditorParams,
      ],
    };

    const record = TestUtils.createPrimitiveStringProperty("Test", "MyValue", "Test", editorInfo);
    const wrapper = mount(<TextEditor propertyRecord={record} />);
    await TestUtils.flushAsyncOperations();

    const textEditor = wrapper.find(TextEditor);
    expect(textEditor.length).to.eq(1);
    expect(textEditor.state("iconSpec")).to.eq(iconSpec);

    wrapper.unmount();
  });

  it("should call onCommit for Enter", async () => {
    const propertyRecord = TestUtils.createPrimitiveStringProperty("Test", "MyValue");
    const convertInfo: PropertyConverterInfo = { name: "" };
    propertyRecord.property.converter = convertInfo;

    const spyOnCommit = sinon.spy();
    function handleCommit(_commit: PropertyUpdatedArgs): void {
      spyOnCommit();
    }
    const wrapper = render(<EditorContainer propertyRecord={propertyRecord} title="abc" onCommit={handleCommit} onCancel={() => { }} />);
    const inputNode = wrapper.container.querySelector("input");
    expect(inputNode).not.to.be.null;

    fireEvent.keyDown(inputNode as HTMLElement, { key: SpecialKey.Enter });
    await TestUtils.flushAsyncOperations();
    expect(spyOnCommit.calledOnce).to.be.true;
  });

  describe("Needs IModelApp", () => {
    before(async () => {
      await TestUtils.initializeUiComponents();
      await MockRender.App.startup();
    });

    after(async () => {
      await MockRender.App.shutdown();
      TestUtils.terminateUiComponents();
    });

    class MineDataController extends DataControllerBase {
      public async validateValue(_newValue: PropertyValue, _record: PropertyRecord): Promise<AsyncValueProcessingResult> {
        return { encounteredError: true, errorMessage: { priority: OutputMessagePriority.Error, briefMessage: "Test"} };
      }
    }

    it("should not commit if DataController fails to validate", async () => {
      PropertyEditorManager.registerDataController("myData", MineDataController);
      const propertyRecord = TestUtils.createPrimitiveStringProperty("Test", "MyValue");
      const convertInfo: PropertyConverterInfo = { name: "" };
      propertyRecord.property.converter = convertInfo;
      propertyRecord.property.dataController = "myData";

      const spyOnCommit = sinon.spy();
      const wrapper = render(<EditorContainer propertyRecord={propertyRecord} title="abc" onCommit={spyOnCommit} onCancel={() => { }} />);
      const inputNode = wrapper.container.querySelector("input");
      expect(inputNode).not.to.be.null;

      fireEvent.keyDown(inputNode as HTMLElement, { key: SpecialKey.Enter });
      await TestUtils.flushAsyncOperations();
      expect(spyOnCommit.calledOnce).to.be.false;

      PropertyEditorManager.deregisterDataController("myData");
    });

    class MineDataController2 extends DataControllerBase {
      public async commitValue(_newValue: PropertyValue, _record: PropertyRecord): Promise<AsyncValueProcessingResult> {
        return { encounteredError: true, errorMessage: { priority: OutputMessagePriority.Error, briefMessage: "Test"} };
      }
    }

    it("should not commit if DataController fails to commit", async () => {
      PropertyEditorManager.registerDataController("myData", MineDataController2);
      const propertyRecord = TestUtils.createPrimitiveStringProperty("Test", "MyValue");
      const convertInfo: PropertyConverterInfo = { name: "" };
      propertyRecord.property.converter = convertInfo;
      propertyRecord.property.dataController = "myData";

      const spyOnCommit = sinon.spy();
      const wrapper = render(<EditorContainer propertyRecord={propertyRecord} title="abc" onCommit={spyOnCommit} onCancel={() => { }} />);
      const inputNode = wrapper.container.querySelector("input");
      expect(inputNode).not.to.be.null;

      fireEvent.keyDown(inputNode as HTMLElement, { key: SpecialKey.Enter });
      await TestUtils.flushAsyncOperations();
      expect(spyOnCommit.calledOnce).to.be.false;

      PropertyEditorManager.deregisterDataController("myData");
    });
  });

});
