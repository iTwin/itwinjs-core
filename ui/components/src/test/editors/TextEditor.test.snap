// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextEditor /> renders correctly 1`] = `
<ForwardRef(Input)
  aria-label="editor.text"
  className="components-cell-editor components-text-editor"
  data-testid="components-text-editor"
  onChange={[Function]}
  readOnly={false}
  style={
    Object {
      "minWidth": "6em",
    }
  }
  type="text"
  value=""
/>
`;

exports[`<TextEditor /> renders correctly with style 1`] = `
<ForwardRef(Input)
  aria-label="editor.text"
  className="components-cell-editor components-text-editor"
  data-testid="components-text-editor"
  onChange={[Function]}
  readOnly={false}
  style={
    Object {
      "color": "red",
    }
  }
  type="text"
  value=""
/>
`;
